src/encryption.rs

```diff
@@ -0,0 +1,1021 @@
+        Ok(EncryptedKey { bytes: BASE64_STANDARD.decode(key.as_ref())? })
+    }
+    pub(crate) fn decrypt_aes_128_ecb(self, decryption_key: &Key) -> std::io::Result<Key> {
+        let cipher = Cipher::aes_128_ecb();
+        if decryption_key.len() != cipher.key_len() {
+            return Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "Invalid key size"));
+        }
+        let bytes = symm::decrypt(cipher, &decryption_key, None, &self.bytes)?;
+
+        Ok(Key { bytes })
+    }
+    pub(crate) fn as_base64(&self) -> String {
+        BASE64_STANDARD.encode(&self.bytes)
+    }
+}
+
+impl Drop for EncryptedKey {
+    fn drop(&mut self) {
+        self.bytes.zeroize();
+    }
+}
+
+pub(crate) struct ContentCryptoMaterial {
+    pub scheme: CryptoScheme,
+    pub cek: Key,
+    pub iv: Iv,
+    pub aad: Option<Bytes>
+}
+
+impl ContentCryptoMaterial {
+    pub fn generate(scheme: CryptoScheme) -> ContentCryptoMaterial {
+        let cek = Key::generate(scheme.key_len());
+        let iv = Iv::generate(scheme.iv_len());
+        ContentCryptoMaterial { scheme, cek, iv, aad: None }
+    }
+
+    pub fn with_aad(self, aad: impl Into<Bytes>) -> ContentCryptoMaterial {
+        ContentCryptoMaterial { aad: Some(aad.into()), ..self }
+    }
+}
+
+pub(crate) fn encrypt(
+    data: &[u8],
+    material: &ContentCryptoMaterial
+) -> std::io::Result<Vec<u8>> {
+    match material.scheme {
+        CryptoScheme::Aes128Cbc => {
+            let ContentCryptoMaterial { scheme, cek, iv, .. } = material;
+            Ok(symm::encrypt(scheme.cipher(), &cek, Some(&iv), data)?)
+
+        },
+        CryptoScheme::Aes256Gcm => {
+            let ContentCryptoMaterial { scheme, cek, iv, aad, .. } = material;
+            let aad_ref = aad.as_deref().unwrap_or_default();
+            let mut tag = vec![0; scheme.tag_len()];
+            let mut ciphertext = encrypt_aead(scheme.cipher(), &cek, Some(&iv), aad_ref, data, &mut tag)?;
+            // Postfix tag
+            ciphertext.extend_from_slice(&tag);
+            Ok(ciphertext)
+        }
+    }
+}
+
+#[allow(unused)]
+pub(crate) fn decrypt(
+    ciphertext: &[u8],
+    material: &ContentCryptoMaterial
+) -> std::io::Result<Vec<u8>> {
+    match material.scheme {
+        CryptoScheme::Aes128Cbc => {
+            let ContentCryptoMaterial { scheme, cek, iv, .. } = material;
+            Ok(symm::decrypt(scheme.cipher(), &cek, Some(&iv), ciphertext)?)
+
+        },
+        CryptoScheme::Aes256Gcm => {
+            let ContentCryptoMaterial { scheme, cek, iv, aad, .. } = material;
+            let aad_ref = aad.as_deref().unwrap_or_default();
+            // Postfix tag
+            let tag_offset = ciphertext.len() - AES_GCM_TAG_BYTES;
+            let data = decrypt_aead(scheme.cipher(), &cek, Some(&iv), aad_ref, &ciphertext[..tag_offset], &ciphertext[tag_offset..])?;
+            Ok(data)
+        }
+    }
+}
+
+struct Buffer {
```

Comment:
For me, a brief description of the core properties of this type would make this easier to read. In particular: what is the special thing here that requires us to introduce this type, rather than using an existing one (I am looking at `BytesMut` for example)

Reply:
The same for some other important types: brief descriptions would help to navigate this code more easily (especially considering that most people at RAI are rust newbies)

Reply:
I agree, we have virtually no description in almost all types at this point and it is a sad state. I will try to address this during the sessions we have planned for knowledge passing.